### PR TITLE
[pigeon] Standardize host api error handling

### PR DESCRIPTION
**Pull request by xegrox**
* [java] Adds a `GeneratedApi.FlutterError` exception for passing custom error parameters (code, message, details).
* [kotlin] Adds a `FlutterError` exception for passing custom error parameters (code, message, details).
* [kotlin] Adds a `errorClassName` option in `KotlinOptions` for custom error class names.
* [java] Removes legacy try catch in async apis.
* Adds `FlutterError` handling integration tests for all platforms.

Fixes flutter/flutter#120861

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

**Comment by tarrinneal** [comment]
> Task :alternate_language_test_plugin:testDebugUnitTest FAILED

Seems there is at least one failing Java test at the moment. 

Thanks for putting this pr together! Let me know if you need any assistance with anything moving forward.

**Comment by xegrox** [comment]
> ! pigeon 9.0.1 from path ../../pigeon (overridden)

What did I do that caused the pigeon version for `webview_flutter` to be overwridden?
It's currently failing because of a regression error

>   error - pigeons/web_kit.dart:19:7 - The named parameter 'header' isn't defined. Try correcting the name to an existing named parameter's name, or defining a named parameter with the name 'header'. - undefined_named_parameter

**Comment by tarrinneal** [comment]
> > ! pigeon 9.0.1 from path ../../pigeon (overridden)
> 
> What did I do that caused the pigeon version for `webview_flutter` to be overwridden? It's currently failing because of a regression error
> 
> > error - pigeons/web_kit.dart:19:7 - The named parameter 'header' isn't defined. Try correcting the name to an existing named parameter's name, or defining a named parameter with the name 'header'. - undefined_named_parameter

Seems unlikely this is from your pr, we just made a major structural change that is most likely causing these issues. I'll look into it to confirm

**Comment by xegrox** [comment]
I noticed this in the swift generator

https://github.com/flutter/packages/blob/c0f0a2250ee0ec0c7bc2323bd24572b2ca3b5a91/packages/pigeon/lib/swift_generator.dart#L474-L476

Is there a reason why the try-catch statement is explicitly ommited for asynchronous methods? Seems to be the case for kotlin as well, compared to the java & cpp generators which retain the try catch statements for async methods.

**Comment by Archifer** [comment]
Hi, I was trying to test this new functionality on the Kotlin side, but ran into a problem; for example, after using the kotlin code generation you could have something like this:

`override fun testFunction(callback: (Result<String>) -> Unit) {
        callback(Result.failure(FlutterException(CODE, MESSAGE, DETAILS)))
    }`

where you want to fail it using a FlutterException. However, initializing the exception is not possible currently because the initializer seems private;  
`Cannot access '<init>': it is package-private in 'FlutterException'`

Should the exception be initiated in a different way for this approach or is there another way how it should be called?

**Comment by xegrox** [comment]
> Hi, I was trying to test this new functionality on the Kotlin side, but ran into a problem; for example, after using the kotlin code generation you could have something like this:
> 
> `override fun testFunction(callback: (Result<String>) -> Unit) { callback(Result.failure(FlutterException(CODE, MESSAGE, DETAILS))) }`
> 
> where you want to fail it using a FlutterException. However, initializing the exception is not possible currently because the initializer seems private; `Cannot access '<init>': it is package-private in 'FlutterException'`
> 
> Should the exception be initiated in a different way for this approach or is there another way how it should be called?

Yes I only realized that recently, it seems that the access modifier for the `FlutterException` constructor may have been missed out. I've created a pr for that at flutter/engine#39817 but if that fails, we could possibly create our own `FlutterException` replacement similar to how it's done in the `cpp_generator`

**Comment by stuartmorgan** [comment]
> I've created a pr for that at [flutter/engine#39817](https://github.com/flutter/engine/pull/39817) but if that fails, we could possibly create our own `FlutterException` replacement similar to how it's done in the `cpp_generator`

What's the argument against making an exception specifically for Pigeon? Only Pigeon code would ever handle it; what would be gained by defining it in the engine?

**Comment by stuartmorgan** [comment]
> Is there a reason why the try-catch statement is explicitly ommited for asynchronous methods? Seems to be the case for kotlin as well, compared to the java & cpp generators which retain the try catch statements for async methods.

Arguably we should remove them for Java and C++. For C++ at least, it's largely a legacy of an earlier error handling design that didn't end up being kept.

The reason to not have them is that it creates two completely different ways to return errors from async methods (one of which will usually not be useful), which is confusing. If a plugin developer is concerned about exceptions in the synchronous part of their async method implementation they can do their own try/catch, just as they would need to in the rest of it.

**Comment by xegrox** [comment]
> > I've created a pr for that at [flutter/engine#39817](https://github.com/flutter/engine/pull/39817) but if that fails, we could possibly create our own `FlutterException` replacement similar to how it's done in the `cpp_generator`
> 
> What's the argument against making an exception specifically for Pigeon? Only Pigeon code would ever handle it; what would be gained by defining it in the engine?

Hmm you are right, it's more sensible to create an exception specifically for pigeon. It's just that the objc/swift side uses `FlutterError`, so I was just trying to follow the same convention.

I'll follow up with a commit to replace `FlutterException` in the kotlin/java side :)

**Comment by stuartmorgan** [comment]
> It's just that the objc/swift side uses `FlutterError`, so I was just trying to follow the same convention.

`FlutterError` is explicitly intended for returning errors over platform channels from native code though; they aren't comparable types in how the underlying engine APIs are designed.

**Comment by xegrox** [comment]
> > It's just that the objc/swift side uses `FlutterError`, so I was just trying to follow the same convention.
> 
> `FlutterError` is explicitly intended for returning errors over platform channels from native code though; they aren't comparable types in how the underlying engine APIs are designed.

Oh I see, I didn't know that and just presumed they were the same. 

Is there a `FlutterError` equivalent for java? One that's for returning errors over platform channels.

**Comment by stuartmorgan** [comment]
In Java values are returned using [Result](https://github.com/flutter/engine/blob/main/shell/platform/android/io/flutter/plugin/common/MethodChannel.java#L194). Everything is async in the engine APIs.

The reason things are more complicated in Pigeon is that it provides a convenience system for the common case of implementations that don't need async handling within the native implementation, so they can return values directly. Language conventions for how to return errors from a synchronous method call differ, and so Pigeon doesn't do the same thing in every language.

**Comment by xegrox** [comment]
I've completed the necessary changes, could someone can review them? Not sure why one of the ci checks timed out though, seems to be unrelated. 


To throw a custom error for Java:
```java
throw GeneratedClass.ApiError("code", "message", "details")
```

Kotlin:
```kotlin
throw MyApiError("code", "message", "details") 
```

Kotlin doesn't have a top level class unlike java, hence I had to provide a different error name for each api to prevent namespace collision.

**Comment by tarrinneal** [comment]
> Kotlin doesn't have a top level class unlike java, hence I had to provide a different error name for each api to prevent namespace collision.

I'm a little confused about why you can't just create a single top level error class for kotlin.

**Comment by xegrox** [comment]
> I'm a little confused about why you can't just create a single top level error class for kotlin. 
> 

Because in java/kotlin, there can't be two classes sharing the same name within the same package

If there are two api files within the same package, the error class will be declared twice even though they are in different files 

ApiOne.kt
```kotlin
package io.flutter

class Test() 
```

ApiTwo.kt
```kotlin
package io.flutter

class Test() // redeclaration error alrdy exists in package io.flutter
```

**Comment by stuartmorgan** [comment]
It's unfortunate to make error handling slightly different in every plugin in Kotlin just for that rare case. What about a `KotlinOptions` entry to allow overriding the generated error class name?

File: packages/pigeon/lib/java_generator.dart [hunk 1]
```diff
@@ -765,22 +771,48 @@ Result<$returnType> $resultName =
...
-\t\t"Cause: " + exception.getCause() + ", Stacktrace: " + Log.getStackTraceString(exception));
+\tif (exception instanceof ApiError) {
+\t\tApiError error = (ApiError) exception;
+\t\terrorList.add(error.code);
+\t\terrorList.add(error.message);
+\t\terrorList.add(error.details);
+\t} else {
+\t\terrorList.add(exception.getClass().getSimpleName());
+\t\terrorList.add(exception.toString());
+\t\terrorList.add(
```
Comment (tarrinneal) [inline packages/pigeon/lib/java_generator.dart:805]: This change makes sense, but I think we should keep the original order of these list items. This would be a breaking change with no "real" benefit.

File: packages/pigeon/lib/swift_generator.dart [hunk 2]
```diff
@@ -630,8 +630,8 @@ import FlutterMacOS
       });
       indent.write('return ');
       indent.addScoped('[', ']', () {
-        indent.writeln(r'"\(error)",');
         indent.writeln(r'"\(type(of: error))",');
+        indent.writeln(r'"\(error)",');
```
Comment (tarrinneal) [inline packages/pigeon/lib/swift_generator.dart:634]: This should also be reverted

File: packages/pigeon/lib/cpp_generator.dart [hunk 3]
```diff
@@ -832,15 +832,15 @@ const flutter::StandardMessageCodec& ${api.name}::GetCodec() {
     indent.format('''
 EncodableValue ${api.name}::WrapError(std::string_view error_message) {
 \treturn EncodableValue(EncodableList{
-\t\tEncodableValue(std::string(error_message)),
 \t\tEncodableValue("Error"),
+\t\tEncodableValue(std::string(error_message)),
```
Comment (tarrinneal) [inline packages/pigeon/lib/cpp_generator.dart:836]: Lets revert this change also.

File: packages/pigeon/platform_tests/test_plugin/windows/test_plugin.cpp [hunk 4]
```diff
@@ -65,7 +65,15 @@ std::optional<FlutterError> TestPlugin::ThrowErrorFromVoid() {
   return FlutterError("An error");
 }
 
-ErrorOr<int64_t> TestPlugin::EchoInt(int64_t an_int) { return an_int; }
+ErrorOr<std::optional<flutter::EncodableValue>> TestPlugin::ThrowFlutterError(){
+    return FlutterError("code", "message", EncodableValue("details"))}
+
+std::optional<FlutterError> TestPlugin::ThrowFlutterErrorFromVoid(){
+    return FlutterError("code", "message", EncodableValue("details"))}
```
Comment (tarrinneal) [inline packages/pigeon/platform_tests/test_plugin/windows/test_plugin.cpp:72]: The formatting on these return lines is incorrect, you'll need a semi-colon after each return (and the `}` should be on it's own line)

**Comment by xegrox** [comment]
> It's unfortunate to make error handling slightly different in every plugin in Kotlin just for that rare case. What about a `KotlinOptions` entry to allow overriding the generated error class name?

Good idea! We could do something similar to `className` in `JavaOptions`; derive the error class name from the kotlin file name by default while also allowing it to be overridden. That way there'll only be one error class per file, similar to java.


> Overall, a great pr. Thank you for putting this together. I have a few nits, and a breaking change I'd prefer reverted.
> 
> If you'd like to discuss the revert I asked for, I'm open to hearing it.
> 
> Good job, and thanks again!

Thanks for the review! I'll push the changes once I've completed the above feature.

**Comment by xegrox** [comment]
To throw a custom error for Java:
```java
throw GeneratedClass.ApiError("code", "message", "details")
```

Kotlin:
```kotlin
throw GeneratedClassError("code", "message", "details") 
```
By default, the error class name for kotlin is derived from the file name, but it can be overridden by the `errorClassName` option in `KotlinOptions`.
  
I also had to replace `path.basenameWithoutExtension` with
```dart
String _basenameWithoutExtensions(String p) {
  return path.basename(p).split('.')[0];
}
```
when deriving the error class name, since `path.basenameWithoutExtensions` only removes the last extension; feeding `CoreTests.gen.kt` would return `CoreTests.gen` instead of `CoreTests`

**Comment by stuartmorgan** [comment]
> By default, the error class name for kotlin is derived from the file name

Why? My suggestion was that we have the default (which is all that the I would expect most clients to ever need) be a generic name so that error handling code is, by default, consistent across plugins rather than always being bespoke.

**Comment by xegrox** [comment]
> Why? My suggestion was that we have the default (which is all that the I would expect most clients to ever need) be a generic name so that error handling code is, by default, consistent across plugins rather than always being bespoke.

Apologies for the misinterpretation, my concern was that the error class names within the same package would conflict; rather than having to set a custom error class name for each api file it would be defaulted to the file name. However I see where you're coming from, it would probably be better to have the error class names be consistent across platforms and plugins.

I've pushed a new commit that sets the default error class name in java/kotlin to `FlutterError` instead, please check if it's alright, thanks!

**Comment by tarrinneal** [comment]
> > Why? My suggestion was that we have the default (which is all that the I would expect most clients to ever need) be a generic name so that error handling code is, by default, consistent across plugins rather than always being bespoke.
> 
> Apologies for the misinterpretation, my concern was that the error class names within the same package would conflict; rather than having to set a custom error class name for each api file it would be defaulted to the file name. However I see where you're coming from, it would probably be better to have the error class names be consistent across platforms and plugins.
> 
> I've pushed a new commit that sets the default error class name in java/kotlin to `FlutterError` instead, please check if it's alright, thanks!

cc @stuartmorgan

**Comment by MichielVrins** [comment]
I have been using this PR and there is actually an inconvenience here for iOS  and Android. 
**iOS/Swift:**  Pigeon creates an interface with callbacks of type `Result<T, Error>` But after this PR it could actually be desirable to solely use `FlutterError` as an expected error. To prevent developer error here, I think it would be useful to have some mechanism to make the created interface have callbacks of type `Result<T, FlutterError>`. Either through an annotation or through a command line argument so we would not introduce breaking changes. 


**Android/Kotlin:** A similar problem arises here as we need to use the `Result` class in kotlin. This results in the same problem as in iOS where you can accidentally pass in the wrong Error. Solving this one is going to be bit more tricky as the Result class does not support typed errors to my knowledge. I suggest that we somehow wrap this Result class plus the suggestion above(in iOS part) of adding a way to force a FlutterError. 

I understand that these changes are bigger than this PR but opening a new issue right now before this got merged seemed wrong.

File: packages/pigeon/CHANGELOG.md [hunk 5]
```diff
@@ -1,10 +1,19 @@
+## 9.0.2
+
+* [java] Adds a `GeneratedApi.FlutterError` exception for passing custom error parameters (code, message, details).
+* [kotlin] Adds a `FlutterError` exception for passing custom error parameters (code, message, details).
+* [kotlin] Adds a `errorClassName` option in `KotlinOptions` for custom error class names.
```
Comment (stuartmorgan) [inline packages/pigeon/CHANGELOG.md:5]: Nit: "Adds an"

File: packages/pigeon/lib/java_generator.dart [hunk 6]
```diff
@@ -606,46 +606,50 @@ class JavaGenerator extends StructuredGenerator<JavaOptions> {
...
                 'ArrayList<Object> wrapped = new ArrayList<Object>();');
-            indent.write('try ');
-            indent.addScoped('{', '}', () {
-              final List<String> methodArgument = <String>[];
-              if (method.arguments.isNotEmpty) {
+            final List<String> methodArgument = <String>[];
+            if (method.arguments.isNotEmpty) {
+              indent.writeln(
+                  'ArrayList<Object> args = (ArrayList<Object>) message;');
+              indent.write('if (args == null) ');
```
Comment (stuartmorgan) [inline packages/pigeon/lib/java_generator.dart:613]: We should not be doing runtime checks that Pigeon-generated messages follow the Pigeon protocol. To the extent that we have them in some languages in some places it's a legacy; we should be removing them rather than adding more.

File: packages/pigeon/lib/java_generator.dart [hunk 7]
```diff
@@ -606,46 +606,50 @@ class JavaGenerator extends StructuredGenerator<JavaOptions> {
...
+                    : argName;
+                String accessor = 'args.get($index)';
+                if (isEnum(root, arg.type)) {
+                  accessor = _intToEnum(accessor, arg.type.baseName);
+                } else if (argType != 'Object') {
+                  accessor = '($argType) $accessor';
+                }
+                indent.writeln('$argType $argName = $accessor;');
+                if (!arg.type.isNullable) {
+                  indent.write('if ($argName == null) ');
```
Comment (stuartmorgan) [inline packages/pigeon/lib/java_generator.dart:638]: Similarly, we can just remove this.

File: packages/pigeon/lib/java_generator.dart [hunk 8]
```diff
@@ -606,46 +606,50 @@ class JavaGenerator extends StructuredGenerator<JavaOptions> {
...
+                  indent.addScoped('{', '}', () {
+                    indent.writeln(
+                        'reply.reply(wrapError(new NullPointerException("aStringArg unexpectedly null.")));');
```
Comment (stuartmorgan) [inline packages/pigeon/lib/java_generator.dart:641]: (If we don't for some reason, this has replaced the variable name with a hard-coded name from one specific output.)

File: packages/pigeon/lib/java_generator.dart [hunk 9]
```diff
@@ -765,22 +771,48 @@ Result<$returnType> $resultName =
...
+      indent.writeln('public final String code;');
+      indent.writeln('public final String message;');
+      indent.writeln('public final Object details;');
+      indent.newln();
+      indent.writeln(
+          'public FlutterError(@NonNull String code, @Nullable String message, @Nullable Object details) ');
+      indent.writeScoped('{', '}', () {
+        indent.writeln('super(message);');
+        indent.writeln('this.code = code;');
+        indent.writeln('this.message = message;');
```
Comment (stuartmorgan) [inline packages/pigeon/lib/java_generator.dart:786]: Why do we need to store the message twice?

File: packages/pigeon/lib/java_generator.dart [hunk 10]
```diff
@@ -765,22 +771,48 @@ Result<$returnType> $resultName =
     });
   }
 
+  void _writeErrorClass(Indent indent) {
+    indent.write('public static class FlutterError extends RuntimeException ');
```
Comment (stuartmorgan) [inline packages/pigeon/lib/java_generator.dart:775]: This class and its public members all need documentation explaining them, per the style guide.

File: packages/pigeon/lib/kotlin_generator.dart [hunk 11]
```diff
@@ -625,25 +629,49 @@ class KotlinGenerator extends StructuredGenerator<KotlinOptions> {
...
+          indent.writeln(
+              '"Cause: " + exception.cause + ", Stacktrace: " + Log.getStackTraceString(exception)');
+        });
       });
     });
   }
 
+  void _writeErrorClass(KotlinOptions generatorOptions, Indent indent) {
+    indent.newln();
+    indent.write('class ${generatorOptions.errorClassName ?? "FlutterError"} ');
```
Comment (stuartmorgan) [inline packages/pigeon/lib/kotlin_generator.dart:660]: Same here; these need comments.

File: packages/pigeon/CHANGELOG.md [hunk 12]
```diff
@@ -1,10 +1,19 @@
+## 9.0.2
+
+* [java] Adds a `GeneratedApi.FlutterError` exception for passing custom error parameters (code, message, details).
+* [kotlin] Adds a `FlutterError` exception for passing custom error parameters (code, message, details).
```
Comment (stuartmorgan) [inline packages/pigeon/CHANGELOG.md:4]: The README should also be updated to explain and demonstrate returning errors with and without these classes.

File: packages/pigeon/pigeons/core_tests.dart [hunk 13]
```diff
@@ -102,9 +102,15 @@ abstract class HostIntegrationCoreApi {
...
 
-  /// Responds with an error from an async void function.
+  /// Returns an error from a void function, to test error handling.
   void throwErrorFromVoid();
 
+  /// Returns a flutter error, to test error handling.
+  Object? throwFlutterError();
+
+  /// Returns a flutter error from a void function, to test error handling.
+  void throwFlutterErrorFromVoid();
```
Comment (stuartmorgan) [inline packages/pigeon/pigeons/core_tests.dart:112]: We don't need two variants of each of the FlutterError versions of the tests, just one. The reason we have two `throw` tests is because the code generated in each case is fundamentally different and so we need to test each, but there's no extra structural difference here that requires two extra tests instead of just one.

File: packages/pigeon/pigeons/core_tests.dart [hunk 14]
```diff
@@ -102,9 +102,15 @@ abstract class HostIntegrationCoreApi {
   /// Returns an error, to test error handling.
   Object? throwError();
 
-  /// Responds with an error from an async void function.
+  /// Returns an error from a void function, to test error handling.
   void throwErrorFromVoid();
 
+  /// Returns a flutter error, to test error handling.
```
Comment (stuartmorgan) [inline packages/pigeon/pigeons/core_tests.dart:108]: Nit: Flutter should be capitalized in all of these comments.

File: packages/pigeon/tool/shared/generation.dart [hunk 15]
```diff
@@ -79,6 +79,7 @@ Future<int> generatePigeons({required String baseDir}) async {
           ? null
           : '$outputBase/android/src/main/kotlin/com/example/test_plugin/$pascalCaseName.gen.kt',
       kotlinPackage: 'com.example.test_plugin',
+      kotlinErrorClassName: '${pascalCaseName}Error',
```
Comment (stuartmorgan) [inline packages/pigeon/tool/shared/generation.dart:82]: I'd like to have end-to-end testing of the default behavior too; let's make the value here `input == 'core_tests' ? null : '${pascalCaseName}Error'`.

**Review by stuartmorgan — CHANGES_REQUESTED** [review]
Some small comments, but overall this looks great!

**Comment by MichielVrins** [comment]
@stuartmorgan I think I indeed might have used the wrong wording in my previous message with too little information to back up my reasoning. 

The Flutter documentation describes that the [PlatformException](https://api.flutter.dev/flutter/services/PlatformException-class.html) has a code (An error code) and can have a Message (A human-readable error message, possibly null) and a stacktrace (Native stacktrace for the error, possibly null).
I would expect this to be the standard for how errors should look like when coming from the native side. But this is currently not the case for any error that is not a FlutterError.

File: packages/pigeon/lib/java_generator.dart (hunk 7 above)
Comment (xegrox) [reply packages/pigeon/lib/java_generator.dart:638]: https://github.com/flutter/packages/blob/b6668e4ef35f3f18373e43df2d699e3da8091dfd/packages/pigeon/platform_tests/alternate_language_test_plugin/android/src/test/java/com/example/alternate_language_test_plugin/PigeonTest.java#L30-L32

This test expects a `NullPointerException` when a argument is unexpectedly null, I assume it should also be removed? Otherwise it fails without the null check.

Comment (stuartmorgan) [reply packages/pigeon/lib/java_generator.dart:638]: Yes, the test should be removed since we're deliberately removing the behavior.

File: packages/pigeon/README.md [hunk 16]
```diff
@@ -224,6 +224,46 @@ abstract class Api2Host {
 }
 ```
 
+### Error Handling
+
+All Host API exceptions are translated into Flutter `PlatformException`.  
+* For synchronous methods, thrown exceptions will be caught and translated.  
+* For asynchronous methods, there is no default exception handling; exceptions should be returned via the provided callback.
```
Comment (stuartmorgan) [inline packages/pigeon/README.md:231]: Nit: "errors should be returned via [...]"

File: packages/pigeon/README.md [hunk 17]
```diff
@@ -224,6 +224,46 @@ abstract class Api2Host {
...
+
+To pass custom details into `PlatformException` for error handling, use `FlutterError` in your Host API instead.
```
Comment (stuartmorgan) [inline packages/pigeon/README.md:233]: Nit: Remove "instead", since it's not being contrasted with anything.

**Review by stuartmorgan — APPROVED** [review]
LGTM with nits, thanks!

I'm in the middle of migrating an Android plugin implementation to Pigeon, and just ran into the need for this in order to preserve the existing error behavior :)

File: packages/pigeon/CHANGELOG.md [hunk 18]
```diff
@@ -29,7 +38,8 @@
 
 ## 9.0.0
 
-* **Breaking Change** Updates `DartOptions` to be immutable and adds const to the constructor.
+* **Breaking Change** Updates `DartOptions` to be immutable and 
+adds const to the constructor.
```
Comment (tarrinneal) [inline packages/pigeon/CHANGELOG.md:42]: This probably wasn't intentional

File: packages/pigeon/CHANGELOG.md [hunk 19]
```diff
@@ -1,3 +1,12 @@
+## 9.0.7
+
+* [java] Adds a `GeneratedApi.FlutterError` exception for passing custom error details (code, message, details).
+* [kotlin] Adds a `FlutterError` exception for passing custom error details (code, message, details).
+* [kotlin] Adds an `errorClassName` option in `KotlinOptions` for custom error class names.
+* [java] Removes legacy try catch from async apis.
+* [java] Removes legacy null check on non-nullable method arguments.
+* Adds `FlutterError` handling integration tests for all platforms.
+
```
Comment (tarrinneal) [inline packages/pigeon/CHANGELOG.md:9]: This list doesn't include the changes you made to the order of items included in a non FlutterError for cpp.

**Review by tarrinneal — CHANGES_REQUESTED** [review]

**Comment by tarrinneal** [comment]
Passes local Android integration tests

**Review by tarrinneal — APPROVED** [review]
Thank you for putting this all together! Really great work here.
